claimtxman/claimtxman.go

```diff
@@ -0,0 +1,411 @@
+		auth:            auth,
+		nonceCache:      cache,
+	}, err
+}
+
+// Start will start the tx management, reading txs from storage,
+// send then to the blockchain and keep monitoring them until they
+// get mined
+func (tm *ClaimTxManager) Start() {
+	for {
+		select {
+		case <-tm.ctx.Done():
+			return
+		case ger := <-tm.chExitRootEvent:
+			err := tm.updateDepositsStatus(ger)
+			if err != nil {
+				log.Errorf("failed to update deposits status: %v", err)
+			}
+		case <-time.After(tm.cfg.FrequencyToMonitorTxs.Duration):
+			err := tm.monitorTxs(context.Background())
+			if err != nil {
+				log.Errorf("failed to monitor txs: %v", err)
+			}
+		}
+	}
+}
+
+func (tm *ClaimTxManager) updateDepositsStatus(ger *etherman.GlobalExitRoot) error {
+	dbTx, err := tm.storage.BeginDBTransaction(tm.ctx)
+	if err != nil {
+		return err
+	}
+	if ger.BlockID != 0 { // L2 exit root is updated
+		log.Infof("Rollup exitroot %v is updated", ger.ExitRoots[1])
+		if err := tm.storage.UpdateL2DepositsStatus(tm.ctx, ger.ExitRoots[1][:], dbTx); err != nil {
+			return err
+		}
+	} else { // L1 exit root is updated in virtual state
+		log.Infof("Mainnet exitroot %v is updated", ger.ExitRoots[0])
+		deposits, err := tm.storage.UpdateL1DepositsStatus(tm.ctx, ger.ExitRoots[0][:], dbTx)
+		if err != nil {
+			return err
+		}
+		for _, deposit := range deposits {
+			claimHash, err := tm.bridgeService.GetDepositStatus(tm.ctx, deposit.DepositCount, deposit.DestinationNetwork)
+			if err != nil {
+				return err
+			}
+			if len(claimHash) > 0 {
+				continue
+			}
+			log.Infof("create the claim tx for the deposit %d", deposit.DepositCount)
+			ger, proves, err := tm.bridgeService.GetClaimProof(deposit.DepositCount, deposit.NetworkID, dbTx)
+			if err != nil {
+				return err
+			}
+			var mtProves [mtHeight][keyLen]byte
+			for i := 0; i < mtHeight; i++ {
+				mtProves[i] = proves[i]
+			}
+			to, data, err := tm.l2Node.BuildSendClaim(tm.ctx, deposit, mtProves,
+				&etherman.GlobalExitRoot{
+					ExitRoots: []common.Hash{
+						ger.ExitRoots[0],
+						ger.ExitRoots[1],
+					}},
+				tm.auth)
+			if err != nil {
+				return err
+			}
+			if err = tm.addClaimTx(deposit.DepositCount, deposit.BlockID, tm.auth.From, to, nil, data, dbTx); err != nil {
+				return err
+			}
+		}
+	}
+	err = tm.storage.Commit(tm.ctx, dbTx)
+	if err != nil {
+		rollbackErr := tm.storage.Rollback(tm.ctx, dbTx)
+		if rollbackErr != nil {
+			log.Fatalf("claimtxman error rolling back state. RollbackErr: %s, err: %s", rollbackErr.Error(), err.Error())
+		}
+		log.Fatalf("AddClaimTx committing dbTx, err: %s", err.Error())
+	}
+	return nil
+}
+
+func (tm *ClaimTxManager) getNextNonce(from common.Address) (uint64, error) {
+	nonce, err := tm.l2Node.NonceAt(tm.ctx, from, nil)
+	if err != nil {
+		return 0, err
+	}
+	if tempNonce, found := tm.nonceCache.Get(from.Hex()); found {
+		if tempNonce >= nonce {
+			nonce = tempNonce + 1
+		}
+	}
+	tm.nonceCache.Add(from.Hex(), nonce)
```

Comment:
because our process is too fast, the nonce is not catching properly, for example through channel, exit_root update events reach up to continuously, in that case, `NonceAt` always returns same nonces (even we have different txs)